File: src/main/java/org/flossware/civilization/model/PopulationState.java
package org.flossware.civilization.model;

/**
 * Immutable population state snapshot.
 */
public record PopulationState(
    long population,
    double birthRate,
    double deathRate,
    double carryingCapacity,
    boolean plagueActive
) {
    public PopulationState {
        if (population < 0) {
            throw new IllegalArgumentException("Population cannot be negative");
        }
        if (carryingCapacity < 0) {
            throw new IllegalArgumentException("Carrying capacity cannot be negative");
        }
    }

    public PopulationState withDelta(double delta) {
        long newPopulation = Math.max(0, population + (long)delta);
        return new PopulationState(
            newPopulation,
            birthRate,
            deathRate,
            carryingCapacity,
            plagueActive
        );
    }

    public PopulationState withPopulation(long newPopulation) {
        return new PopulationState(newPopulation, birthRate, deathRate, carryingCapacity, plagueActive);
    }

    public PopulationState withCarryingCapacity(double newCapacity) {
        return new PopulationState(population, birthRate, deathRate, newCapacity, plagueActive);
    }

    public PopulationState withPlague(boolean isActive) {
        return new PopulationState(population, birthRate, deathRate, carryingCapacity, isActive);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/PopulationState.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable

@Serializable
data class PopulationState(
    val population: Long,
    val birthRate: Double,
    val deathRate: Double,
    val carryingCapacity: Double,
    val plagueActive: Boolean
) {
    init {
        require(population >= 0) { "Population cannot be negative" }
        require(carryingCapacity >= 0) { "Carrying capacity cannot be negative" }
    }

    fun withDelta(delta: Double): PopulationState {
        val newPopulation = maxOf(0L, population + delta.toLong())
        return copy(population = newPopulation)
    }

    fun withPopulation(newPopulation: Long) = copy(population = newPopulation)
    fun withCarryingCapacity(newCapacity: Double) = copy(carryingCapacity = newCapacity)
    fun withPlague(isActive: Boolean) = copy(plagueActive = isActive)
}
